File: src/mcf/lha.cpp
// LHA static-Huffman (-lh5-) decoder.
//
// Identified from uncompress_int()'s allocation fingerprint in libmcfandroid.so
// (NC=510, NT=19, u16[2*NC-1] tree arrays, c_table[4096]/pt_table[256], and five
// ring buffers of 2^11..2^15). See docs/mpk-format.md. Mirrors tools/asset/lha.py.
#include "mcf/mcf.h"

#include <array>
#include <cstring>

namespace mcf::lha {
namespace {

constexpr int kUCharMax = 255;
constexpr int kThreshold = 3;
constexpr int kNC = 510;
constexpr int kNT = 19;
constexpr int kCBit = 9;
constexpr int kTBit = 5;

class BitReader {
public:
    explicit BitReader(std::span<const uint8_t> d) : d_(d) { Fill(16); }

    void Fill(int n) {
        bitbuf_ = uint16_t(bitbuf_ << n);
        while (n > bitcount_) {
            n -= bitcount_;
            bitbuf_ |= uint16_t(subbitbuf_ << n);
            subbitbuf_ = p_ < d_.size() ? d_[p_++] : 0;
            bitcount_ = 8;
        }
        bitcount_ -= n;
        bitbuf_ |= uint16_t(subbitbuf_ >> bitcount_);
    }

    uint32_t Get(int n) {
        if (n == 0) return 0;
        uint32_t x = uint32_t(bitbuf_) >> (16 - n);
        Fill(n);
        return x;
    }

    uint16_t buf() const { return bitbuf_; }
    size_t consumed() const { return p_; }

private:
    std::span<const uint8_t> d_;
    size_t p_ = 0;
    uint16_t bitbuf_ = 0;
    uint8_t subbitbuf_ = 0;
    int bitcount_ = 0;
};

class Decoder {
public:
    Decoder(std::span<const uint8_t> in, int dicbit)
        : br_(in), dicsiz_(1u << dicbit), np_(dicbit + 1),
          pbit_(dicbit <= 13 ? 4 : 5), npt_(std::max(kNT, np_)) {
        pt_len_.resize(npt_);
    }

    void Run(std::span<uint8_t> out) {
        std::vector<uint8_t> ring(dicsiz_);
        uint32_t r = 0;
        size_t n = 0;
        while (n < out.size()) {
            int c = DecodeC();
            if (c <= kUCharMax) {
                ring[r] = uint8_t(c);
                out[n++] = uint8_t(c);
                r = (r + 1) & (dicsiz_ - 1);
            } else {
                int len = c - (kUCharMax + 1 - kThreshold);
                uint32_t i = (r - DecodeP() - 1) & (dicsiz_ - 1);
                for (int k = 0; k < len && n < out.size(); ++k) {
                    uint8_t b = ring[i];
                    ring[r] = b;
                    out[n++] = b;
                    i = (i + 1) & (dicsiz_ - 1);
                    r = (r + 1) & (dicsiz_ - 1);
                }
            }
        }
    }

    size_t consumed() const { return br_.consumed(); }

private:
    // Canonical LHA make_table. `table` is the 2^tablebits fast lookup; codes
    // longer than tablebits spill into the left_/right_ binary tree.
    void MakeTable(int nchar, const std::vector<uint8_t>& bitlen, int tablebits,
                   std::vector<uint16_t>& table) {
        std::array<int, 17> count{}, weight{};
        std::array<int, 18> start{};
        for (int i = 0; i < nchar; ++i) count[bitlen[i]]++;
        for (int i = 1; i < 17; ++i) start[i + 1] = start[i] + (count[i] << (16 - i));
        if (start[17] != 0x10000) throw Error("LHA: malformed Huffman table");

        int jutbits = 16 - tablebits;
        for (int i = 1; i <= tablebits; ++i) {
            start[i] >>= jutbits;
            weight[i] = 1 << (tablebits - i);
        }
        for (int i = tablebits + 1; i < 17; ++i) weight[i] = 1 << (16 - i);

        int i = start[tablebits + 1] >> jutbits;
        if (i != 0)
            for (int k = i; k < (1 << tablebits); ++k) table[k] = 0;

        int avail = nchar;
        int mask = 1 << (15 - tablebits);
        for (int ch = 0; ch < nchar; ++ch) {
            int ln = bitlen[ch];
            if (ln == 0) continue;
            int nextcode = start[ln] + weight[ln];
            if (ln <= tablebits) {
                for (int k = start[ln]; k < std::min(nextcode, 1 << tablebits); ++k)
                    table[k] = uint16_t(ch);
            } else {
                int k = start[ln];
                uint16_t* p = &table[k >> jutbits];
                for (int b = 0; b < ln - tablebits; ++b) {
                    if (*p == 0) {
                        right_[avail] = left_[avail] = 0;
                        *p = uint16_t(avail++);
                    }
                    p = (k & mask) ? &right_[*p] : &left_[*p];
                    k = uint16_t(k << 1);
                }
                *p = uint16_t(ch);
            }
            start[ln] = nextcode;
        }
    }

    void ReadPtLen(int nn, int nbit, int i_special) {
        int n = int(br_.Get(nbit));
        if (n == 0) {
            uint16_t c = uint16_t(br_.Get(nbit));
            std::fill(pt_len_.begin(), pt_len_.begin() + nn, uint8_t(0));
            std::fill(pt_table_.begin(), pt_table_.end(), c);
            return;
        }
        int i = 0;
        while (i < n) {
            int c = br_.buf() >> 13;
            if (c == 7) {
                int m = 1 << 12;
                while (m & br_.buf()) { m >>= 1; ++c; }
            }
            br_.Fill(c < 7 ? 3 : c - 3);
            pt_len_[i++] = uint8_t(c);
            if (i == i_special) {
                int z = int(br_.Get(2));
                while (z-- > 0 && i < nn) pt_len_[i++] = 0;
            }
        }
        while (i < nn) pt_len_[i++] = 0;
        MakeTable(nn, pt_len_, 8, pt_table_);
    }

    void ReadCLen() {
        int n = int(br_.Get(kCBit));
        if (n == 0) {
            uint16_t c = uint16_t(br_.Get(kCBit));
            std::fill(c_len_.begin(), c_len_.end(), uint8_t(0));
            std::fill(c_table_.begin(), c_table_.end(), c);
            return;
        }
        int i = 0;
        while (i < n) {
            int c = pt_table_[br_.buf() >> 8];
            if (c >= kNT) {
                int m = 1 << 7;
                do {
                    c = (br_.buf() & m) ? right_[c] : left_[c];
                    m >>= 1;
                } while (c >= kNT);
            }
            br_.Fill(pt_len_[c]);
            if (c <= 2) {
                if (c == 0) c = 1;
                else if (c == 1) c = int(br_.Get(4)) + 3;
                else c = int(br_.Get(kCBit)) + 20;
                while (c-- > 0) c_len_[i++] = 0;
            } else {
                c_len_[i++] = uint8_t(c - 2);
            }
        }
        while (i < kNC) c_len_[i++] = 0;
        MakeTable(kNC, c_len_, 12, c_table_);
    }

    int DecodeC() {
        if (blocksize_ == 0) {
            blocksize_ = int(br_.Get(16));
            ReadPtLen(kNT, kTBit, 3);
            ReadCLen();
            ReadPtLen(np_, pbit_, -1);
        }
        --blocksize_;
        int j = c_table_[br_.buf() >> 4];
        if (j >= kNC) {
            int m = 1 << 3;
            do {
                j = (br_.buf() & m) ? right_[j] : left_[j];
                m >>= 1;
            } while (j >= kNC);
        }
        br_.Fill(c_len_[j]);
        return j;
    }

    uint32_t DecodeP() {
        int j = pt_table_[br_.buf() >> 8];
        if (j >= np_) {
            int m = 1 << 7;
            do {
                j = (br_.buf() & m) ? right_[j] : left_[j];
                m >>= 1;
            } while (j >= np_);
        }
        br_.Fill(pt_len_[j]);
        if (j == 0) return 0;
        return (1u << (j - 1)) + br_.Get(j - 1);
    }

    BitReader br_;
    uint32_t dicsiz_;
    int np_, pbit_, npt_;
    int blocksize_ = 0;
    std::vector<uint8_t> c_len_ = std::vector<uint8_t>(kNC);
    std::vector<uint8_t> pt_len_;
    std::vector<uint16_t> c_table_ = std::vector<uint16_t>(4096);
    std::vector<uint16_t> pt_table_ = std::vector<uint16_t>(256);
    std::vector<uint16_t> left_ = std::vector<uint16_t>(2 * kNC - 1);
    std::vector<uint16_t> right_ = std::vector<uint16_t>(2 * kNC - 1);
};

}  // namespace

void Decode(std::span<const uint8_t> in, std::span<uint8_t> out, size_t* consumed,
            int dicbit) {
    Decoder d(in, dicbit);
    d.Run(out);
    if (consumed) *consumed = d.consumed();
}

}  // namespace mcf::lha

File: tools/asset/lha.py
"""LHA static-Huffman (-lh*-) decoder, as used by libmcfandroid's uncompress().

Identified from uncompress_int()'s allocation fingerprint: NC=510, NT=19,
left/right as u16[2*NC-1], c_table[4096]/pt_table[256], and five ring buffers
of 2^11..2^15 bytes. See docs/mpk-format.md.
"""
UCHAR_MAX, THRESHOLD, NC, NT, CBIT, TBIT = 255, 3, 510, 19, 9, 5


class BitReader:
    def __init__(self, data):
        self.d, self.p = data, 0
        self.bitbuf = self.subbitbuf = self.bitcount = 0
        self.fillbuf(16)

    def _getc(self):
        if self.p < len(self.d):
            c = self.d[self.p]; self.p += 1; return c
        return 0

    def fillbuf(self, n):
        self.bitbuf = (self.bitbuf << n) & 0xFFFF
        while n > self.bitcount:
            n -= self.bitcount
            self.bitbuf |= (self.subbitbuf << n) & 0xFFFF
            self.subbitbuf = self._getc()
            self.bitcount = 8
        self.bitcount -= n
        self.bitbuf |= self.subbitbuf >> self.bitcount

    def getbits(self, n):
        if n == 0:
            return 0
        x = self.bitbuf >> (16 - n)
        self.fillbuf(n)
        return x


class Decoder:
    def __init__(self, data, dicbit):
        self.br = BitReader(data)
        self.dicbit = dicbit
        self.dicsiz = 1 << dicbit
        self.np = dicbit + 1
        self.pbit = 4 if dicbit <= 13 else 5
        self.npt = max(NT, self.np)
        self.c_len = bytearray(NC)
        self.pt_len = bytearray(self.npt)
        self.c_table = [0] * 4096
        self.pt_table = [0] * 256
        self.left = [0] * (2 * NC - 1)
        self.right = [0] * (2 * NC - 1)
        self.blocksize = 0

    # --- canonical LHA make_table -------------------------------------------
    def make_table(self, nchar, bitlen, tablebits, table):
        count = [0] * 17
        weight = [0] * 17
        start = [0] * 18
        for i in range(nchar):
            count[bitlen[i]] += 1
        for i in range(1, 17):
            start[i + 1] = start[i] + (count[i] << (16 - i))
        if start[17] != 0x10000:
            raise ValueError("bad Huffman table (start[17]=%#x)" % start[17])
        jutbits = 16 - tablebits
        for i in range(1, tablebits + 1):
            start[i] >>= jutbits
            weight[i] = 1 << (tablebits - i)
        for i in range(tablebits + 1, 17):
            weight[i] = 1 << (16 - i)
        i = start[tablebits + 1] >> jutbits
        if i != 0:
            for k in range(i, 1 << tablebits):
                table[k] = 0
        avail = nchar
        mask = 1 << (15 - tablebits)
        for ch in range(nchar):
            ln = bitlen[ch]
            if ln == 0:
                continue
            nextcode = start[ln] + weight[ln]
            if ln <= tablebits:
                for i in range(start[ln], min(nextcode, 1 << tablebits)):
                    table[i] = ch
            else:
                k = start[ln]
                idx = k >> jutbits
                pt, pi = table, idx          # (container, index) cursor
                for _ in range(ln - tablebits):
                    if pt[pi] == 0:
                        self.right[avail] = self.left[avail] = 0
                        pt[pi] = avail
                        avail += 1
                    pt, pi = ((self.right, pt[pi]) if (k & mask)
                              else (self.left, pt[pi]))
                    k = (k << 1) & 0xFFFF
                pt[pi] = ch
            start[ln] = nextcode

    def read_pt_len(self, nn, nbit, i_special):
        br = self.br
        n = br.getbits(nbit)
        if n == 0:
            c = br.getbits(nbit)
            for i in range(nn):
                self.pt_len[i] = 0
            for i in range(256):
                self.pt_table[i] = c
            return
        i = 0
        while i < n:
            c = br.bitbuf >> 13
            if c == 7:
                mask = 1 << 12
                while mask & br.bitbuf:
                    mask >>= 1
                    c += 1
            br.fillbuf(3 if c < 7 else c - 3)
            self.pt_len[i] = c
            i += 1
            if i == i_special:
                c = br.getbits(2)
                while c > 0 and i < nn:
                    self.pt_len[i] = 0
                    i += 1
                    c -= 1
        while i < nn:
            self.pt_len[i] = 0
            i += 1
        self.make_table(nn, self.pt_len, 8, self.pt_table)

    def read_c_len(self):
        br = self.br
        n = br.getbits(CBIT)
        if n == 0:
            c = br.getbits(CBIT)
            for i in range(NC):
                self.c_len[i] = 0
            for i in range(4096):
                self.c_table[i] = c
            return
        i = 0
        while i < n:
            c = self.pt_table[br.bitbuf >> 8]
            if c >= NT:
                mask = 1 << 7
                while True:
                    c = self.right[c] if (br.bitbuf & mask) else self.left[c]
                    mask >>= 1
                    if c < NT:
                        break
            br.fillbuf(self.pt_len[c])
            if c <= 2:
                if c == 0:
                    c = 1
                elif c == 1:
                    c = br.getbits(4) + 3
                else:
                    c = br.getbits(CBIT) + 20
                while c > 0:
                    self.c_len[i] = 0
                    i += 1
                    c -= 1
            else:
                self.c_len[i] = c - 2
                i += 1
        while i < NC:
            self.c_len[i] = 0
            i += 1
        self.make_table(NC, self.c_len, 12, self.c_table)

    def decode_c(self):
        br = self.br
        if self.blocksize == 0:
            self.blocksize = br.getbits(16)
            self.read_pt_len(NT, TBIT, 3)
            self.read_c_len()
            self.read_pt_len(self.np, self.pbit, -1)
        self.blocksize -= 1
        j = self.c_table[br.bitbuf >> 4]
        if j >= NC:
            mask = 1 << 3
            while True:
                j = self.right[j] if (br.bitbuf & mask) else self.left[j]
                mask >>= 1
                if j < NC:
                    break
        br.fillbuf(self.c_len[j])
        return j

    def decode_p(self):
        br = self.br
        j = self.pt_table[br.bitbuf >> 8]
        if j >= self.np:
            mask = 1 << 7
            while True:
                j = self.right[j] if (br.bitbuf & mask) else self.left[j]
                mask >>= 1
                if j < self.np:
                    break
        br.fillbuf(self.pt_len[j])
        if j != 0:
            j = (1 << (j - 1)) + br.getbits(j - 1)
        return j

    def decode(self, outsize):
        out = bytearray(outsize)
        buf = bytearray(self.dicsiz)
        r = 0
        n = 0
        while n < outsize:
            c = self.decode_c()
            if c <= UCHAR_MAX:
                buf[r] = c
                out[n] = c
                n += 1
                r = (r + 1) & (self.dicsiz - 1)
            else:
                j = c - (UCHAR_MAX + 1 - THRESHOLD)
                i = (r - self.decode_p() - 1) & (self.dicsiz - 1)
                for _ in range(j):
                    if n >= outsize:
                        break
                    b = buf[i]
                    buf[r] = b
                    out[n] = b
                    n += 1
                    i = (i + 1) & (self.dicsiz - 1)
                    r = (r + 1) & (self.dicsiz - 1)
        return bytes(out)


def decompress(data, outsize, dicbit):
    return Decoder(data, dicbit).decode(outsize)
